File: highRisen/logic.java
import java.io.*;
import java.util.*;
import java.text.*;
import java.math.*;
import java.util.regex.*;

public class Solution {

    public static void main(String[] args) {
        /* Enter your code here. Read input from STDIN. Print output to STDOUT. Your class should be named Solution. */
        Scanner in=new Scanner(System.in);
        int t=in.nextInt();
        while(t>0)
        {
            int r=in.nextInt();
            int n=in.nextInt();
            int blocks=(int)(Math.log(r)/Math.log(4));
            blocks--;
            if(blocks<=2)
                System.out.println("NO SHIFTING");
            else
            {
                long x=0, min=0;
                x=(long)((-1+Math.sqrt(8*n-7))/2);
                min=1+(long)((n-(x*(x-1))/2-1)*Math.pow(2,x));
                System.out.println(min%1000000007);
            }
        }
    }
}

File: highRisen/logic.cpp
#include <bits/stdc++.h>

using namespace std;

int main(){
    cin.tie(NULL);
    cout.tie(NULL);
    ios_base::sync_with_stdio(false);
        int T;
        cin>>T;
        while(T--){
                int r,blocks;
                long long n;
                cin>>r>>n;
                blocks = floor( log(r)/log(4) ) - 1;
                if( blocks == 1 || blocks == 2 ){
                    cout<<"NO SHIFTING\n";
                }
                else{
                    long long x, min_steps = 0;
                    x = floor( ( -1 + (sqrt( 8*n - 7 )) )/2 );
                    min_steps = 1 + floor( n - (x*(x - 1))/2 - 1)*pow(2,x);
                    cout<<min_steps%(1000000007)<<"\n";
                }
        }
        return 0;
}